File: src/main/java/com/framework/pages/CartPage.java
package com.framework.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Page Object for the SauceDemo Cart page.
 */
public class CartPage extends BasePage {

    // ===== Locators (XPath) =====
    private final By pageTitle = By.xpath("//span[@class='title']");
    private final By cartItems = By.xpath("//div[@class='cart_item']");
    private final By cartItemNames = By.xpath("//div[@class='inventory_item_name']");
    private final By removeButtons = By.xpath("//button[contains(@id,'remove')]");
    private final By checkoutButton = By.xpath("//button[@id='checkout']");
    private final By continueShoppingButton = By.xpath("//button[@id='continue-shopping']");

    public CartPage(WebDriver driver) {
        super(driver);
    }

    // ===== Page Actions =====

    public String getPageTitleText() {
        return getText(pageTitle);
    }

    public boolean isCartPageDisplayed() {
        return isDisplayed(pageTitle);
    }

    public int getCartItemCount() {
        List<WebElement> items = driver.findElements(cartItems);
        logger.info("Cart items count: {}", items.size());
        return items.size();
    }

    public List<String> getCartItemNames() {
        return driver.findElements(cartItemNames).stream()
                .map(WebElement::getText)
                .collect(Collectors.toList());
    }

    public void removeFirstItem() {
        logger.info("Removing first item from cart");
        List<WebElement> buttons = driver.findElements(removeButtons);
        if (!buttons.isEmpty()) {
            buttons.get(0).click();
        }
    }

    public void clickCheckout() {
        logger.info("Clicking checkout button");
        click(checkoutButton);
    }

    public void clickContinueShopping() {
        logger.info("Clicking continue shopping");
        click(continueShoppingButton);
    }
}

File: pages/cart_page.py
"""Page Object for the Sauce Demo Cart page."""

from pages.base_page import BasePage
from utils.logger import get_logger

logger = get_logger("cart_page")


class CartPage(BasePage):
    """Encapsulates locators and actions for the shopping cart page."""

    # ── XPath Locators ───────────────────────────────────────
    PAGE_TITLE = "//span[@class='title']"
    CART_ITEM = "//div[@class='cart_item']"
    CART_ITEM_NAME = "//div[@class='inventory_item_name']"
    CHECKOUT_BUTTON = "//button[@id='checkout']"
    CONTINUE_SHOPPING_BUTTON = "//button[@id='continue-shopping']"

    @staticmethod
    def remove_item_button(item_name: str) -> str:
        return (
            f"//div[text()='{item_name}']/ancestor::div[@class='cart_item']"
            f"//button[contains(@id,'remove')]"
        )

    # ── Page actions ─────────────────────────────────────────
    def get_page_title(self) -> str:
        return self.get_text(self.PAGE_TITLE)

    def get_cart_items_count(self) -> int:
        return self.get_element_count(self.CART_ITEM)

    def get_cart_item_names(self) -> list[str]:
        elements = self.page.locator(self.CART_ITEM_NAME)
        return [elements.nth(i).inner_text() for i in range(elements.count())]

    def remove_item(self, item_name: str):
        logger.info("Removing item from cart: %s", item_name)
        self.click(self.remove_item_button(item_name), step_name="remove_cart_item")
        return self

    def click_checkout(self):
        self.click(self.CHECKOUT_BUTTON, step_name="click_checkout")
        return self

    def click_continue_shopping(self):
        self.click(self.CONTINUE_SHOPPING_BUTTON, step_name="continue_shopping")
        return self
